src/business-logic/services/rundown-lock-service.ts

```diff
@@ -0,0 +1,125 @@
+import { Part } from '../../model/entities/part'
+import { Piece } from '../../model/entities/piece'
+import { Owner } from '../../model/enums/owner'
+import { InTransition } from '../../model/value-objects/in-transition'
+import { RundownService } from './interfaces/rundown-service'
+
+const RUNDOWN_LOCK_INTERVAL = 500
+const RUNDOWN_LOCK_ERROR_TEXT = 'Unable to do action. An action was already executed less than 500ms ago'
+
+export class RundownLockService implements RundownService {
+  private static instance: RundownService
+  private hasEnoughTimePassed: boolean = true
```

Comment:
You don't need a boolean for this (nor be using a `timeout` which you call in every function).

You can simply store the timestamp of the last executed action and then, when you call an action, you check if `Date.now()` is greater than `timestamp + lock_delay`.
It will look something like this:

```
RundownLockService ... {
  private lastActionCalledTimestamp: number

  public takeNext(...): ... {
    this.assertEnoughTimeHasPassed()
    this.rundownService.takeNext(...)
  }

  private assertEnoughTimeHasPassed(): void {
    const now: number = Date.now()
    if (now < (this.lastActionCalledTimestamp + RUNDOWN_LOCK_INTERVAL_MS) {
      throw new RundownIsLockedException('...')
    }
    this.lastActionCalledTimestamp = now
  }
}
```

Then all you have to in each method that needs to be locked, is to add the line `this.assertEnoughTimeHasPassed()`

Reply:
I rework the logic to be that way